test/helpers/kubectl.go

```diff
@@ -2916,36 +2917,64 @@ func (kub *Kubectl) GetCiliumPodOnNodeWithLabel(namespace string, label string)
 	return kub.GetCiliumPodOnNode(namespace, node)
 }
 
-func (kub *Kubectl) ciliumPreFlightCheck() error {
-	err := kub.ciliumStatusPreFlightCheck()
-	if err != nil {
-		return fmt.Errorf("status is unhealthy: %s", err)
-	}
+func (kub *Kubectl) validateCilium() error {
+	var (
+		wg     sync.WaitGroup
+		errors = make(chan error, 128)
+	)
 
-	err = kub.ciliumControllersPreFlightCheck()
-	if err != nil {
-		return fmt.Errorf("controllers are failing: %s", err)
-	}
+	wg.Add(1)
+	go func() {
+		if err := kub.ciliumStatusPreFlightCheck(); err != nil {
+			errors <- fmt.Errorf("status is unhealthy: %s", err)
+		}
+		wg.Done()
+	}()
+
+	wg.Add(1)
+	go func() {
+		if err := kub.ciliumControllersPreFlightCheck(); err != nil {
+			errors <- fmt.Errorf("controllers are failing: %s", err)
+		}
+		wg.Done()
+	}()
 
 	switch GetCurrentIntegration() {
 	case CIIntegrationFlannel:
 	default:
-		err = kub.ciliumHealthPreFlightCheck()
+		wg.Add(1)
+		go func() {
+			if err := kub.ciliumHealthPreFlightCheck(); err != nil {
+				errors <- fmt.Errorf("connectivity health is failing: %s", err)
+			}
+			wg.Done()
+		}()
+	}
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		err := kub.fillServiceCache()
 		if err != nil {
-			return fmt.Errorf("connectivity health is failing: %s", err)
+			errors <- fmt.Errorf("unable to fill service cache: %s", err)
+			return
 		}
-	}
-	err = kub.fillServiceCache()
-	if err != nil {
-		return fmt.Errorf("unable to fill service cache: %s", err)
-	}
-	err = kub.ciliumServicePreFlightCheck()
-	if err != nil {
-		return fmt.Errorf("cilium services are not set up correctly: %s", err)
-	}
-	err = kub.servicePreFlightCheck("kubernetes", "default")
-	if err != nil {
-		return fmt.Errorf("kubernetes service is not ready: %s", err)
+		err = kub.ciliumServicePreFlightCheck()
+		if err != nil {
+			errors <- fmt.Errorf("cilium services are not set up correctly: %s", err)
+			return
+		}
+		err = kub.servicePreFlightCheck("kubernetes", "default")
+		if err != nil {
+			errors <- fmt.Errorf("kubernetes service is not ready: %s", err)
+		}
+	}()
+
+	wg.Wait()
+	select {
+	case err := <-errors:
```

Comment:
Shouldn't this be `for err := range errors { }`?

Reply:
That's because we are not closing the channel. Should it be closed in a separate goroutine waiting for a waitgroup?

Reply:
If you use `x/sync/errgroup`, as suggested above, you can simply use `err := wg.Wait()` as it returns the first non-nil error returned by one of the group's goroutine (if any).

Reply:
Seems like an odd choice to keep the channel open, closing it would result in more implicit control flow and thereby more readable code. I agree it can get a little (tedious)(https://golangcode.com/errors-in-waitgroups/), but hopefully with `errgroup` you'd be able to tidy up this code more generally.

Another reason I wouldn't use `select` here is because it only catches one error and forgets the rest, or was that also somehow intentional?

Reply:
If all you wanted is just one error, you don't need a select, just read and return it if not nil.